### PR TITLE
Various Bugfixes in the Lecture Unit Feature

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [X] Server: I implemented the changes with a good performance and prevented too many database calls

### Motivation and Context

The Lecture Unit Feature introduced several small bugs and inconstancies that need to be fixed before the lecture period starts on Monday November 2nd 2020.

### Description

Commit 3ecd84d7dc75d68715a44da8436bb505b1f9a4cc re-activates release date locking for regular lecture attachments.

Commit 450d9cb84e0cd7202bfbb545522eb8684be69595 ef450c2898427ae905fa27993d8fd72040e5bd8e fixed package structure not being equal to the folder structure

Commit 83e458521c10d309ae7fa652992da870a08bd18b removes unnecessary @Cache annotation on subclasses of LectureUnit

Commit b8fc707bc77bec64004069e1117a3501d7078931 **makes exercise units behave exactly in the same way as in the course dashboard**. NOTE: _The implementation is rather naive and just uses the underlying for-dashboard call. This will be optimized in a follow up, but I think this quick is the most appropriate right now as the for-dashboard call is tested and covers all edge cases like quizzes and so on._

Commit 80527dd9b4e99b16170bae32619ec542f89cd22a is just refactoring to make it more clearer that these are exclusive options.

### Steps for Testing

- Test that the bugs described in the Description section are fixed

### Test Coverage
No changes as just bugfixes

### Screenshots
No changes as just bugfixes

### CODE REVIEW

**Review by MaisiKoleni — COMMENTED** [review]
Do we have any documentation that `Lecture` and `AttachmentUnit` are exclusive? At least when reviewing the changes this is not clear to me. Otherwise, the changes took fine to me and the packages work now (i.e. the code now compiles in Eclipse again).

File: src/main/java/de/tum/in/www1/artemis/domain/Attachment.java [hunk 1]
```diff
@@ -83,7 +83,7 @@ public void onLoad() {
         if (attachmentType == AttachmentType.FILE && getLecture() != null && link != null && link.contains(Constants.FILEPATH_ID_PLACHEOLDER)) {
             link = link.replace(Constants.FILEPATH_ID_PLACHEOLDER, getLecture().getId().toString());
         }
-        if (attachmentType == AttachmentType.FILE && getAttachmentUnit() != null && link != null && link.contains(Constants.FILEPATH_ID_PLACHEOLDER)) {
+        else if (attachmentType == AttachmentType.FILE && getAttachmentUnit() != null && link != null && link.contains(Constants.FILEPATH_ID_PLACHEOLDER)) {
             link = link.replace(Constants.FILEPATH_ID_PLACHEOLDER, getAttachmentUnit().getId().toString());
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/domain/Attachment.java:87]: `Constants.FILEPATH_ID_PLACHEOLDER` has a typo. Maybe we can fix it with your PR?

**Review by derLalla — DISMISSED** [review]
I have not tested it, but the code looks really good to me!

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/lectureUnit.service.ts [hunk 2]
```diff
@@ -21,6 +21,14 @@ export class LectureUnitService {
     constructor(private httpClient: HttpClient, private attachmentService: AttachmentService, private exerciseService: ExerciseService) {}
 
     updateOrder(lectureId: number, lectureUnits: LectureUnit[]): Observable<HttpResponse<LectureUnit[]>> {
+        // need to remove participations from exercise units to prevent circular structure failure
+        lectureUnits = lectureUnits.map((lectureUnit) => {
+            if (lectureUnit.type === LectureUnitType.EXERCISE && (lectureUnit as ExerciseUnit).exercise) {
+                (lectureUnit as ExerciseUnit).exercise!.studentParticipations = undefined;
+                (lectureUnit as ExerciseUnit).exercise!.tutorParticipations = undefined;
+            }
```
Comment (FrankeLukas) [inline src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/lectureUnit.service.ts:29]: Casting three times doesn't look as performant as casting once, so I suggest doing something like:
```
if (lectureUnit.type !== LectureUnitType.EXERCISE) {
    return lectureUnit;
}
const exercise = (lectureUnit as ExerciseUnit).exercise;
exercise!.studentParticipations = undefinded;
exercise!.tutorParticipations = undefined;
return lectureUnit;
```

**Review by FrankeLukas — DISMISSED** [review]
I clicked through the lecture units in your test course on TS1 and didn't find any issues.

File: src/main/java/de/tum/in/www1/artemis/domain/Attachment.java (hunk 1 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/Attachment.java]: Done in 9a0a229d1accaf5134471f4873deb07e83db44df

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/lectureUnit.service.ts (hunk 2 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/lectureUnit.service.ts:29]: Do not agree the new variable is imo more of a performance hit than giving the compiler type assertions

Comment (FrankeLukas) [reply src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/lectureUnit.service.ts:29]: I see, that is just compile time casting, so you're right.

**Review by krusche — APPROVED** [review]

**Pull request merged**